scripts/precomputation.py

```diff
@@ -0,0 +1,301 @@
+
+print("TorchANI version: ", torchani.__version__)
+
+from typing import Optional, Tuple
+
+from torch import Tensor
+from torchani.nn import ANIModel, SpeciesEnergies
+
+torch.set_num_threads(1)
+
+
+class PartialANIModel(ANIModel):
+    """just like ANIModel, but don't do the sum over atoms in the last step, and
+    don't flatten last layer output!"""
+
+    def forward(
+        self,
+        species_aev: Tuple[Tensor, Tensor],  # type: ignore
+        cell: Optional[Tensor] = None,
+        pbc: Optional[Tensor] = None,
+    ) -> Tuple[Tensor, Tensor]:
+
+        species, aev = species_aev
+        assert species.shape == aev.shape[:-1]
+
+        # in our case, species will be the same for all snapshots
+        atom_species = species[0]
+        assert (atom_species == species).all()
+
+        # NOTE: depending on the element, outputs will have different dimensions...
+        # something like output.shape = n_snapshots, n_atoms, n_dims
+        # where n_dims is either 160, 128, or 96...
+
+        # Ugly hard-coding approach: make this of size max_dim=160 and only write
+        # into the first 96, 128, or 160 elements, NaN-poisoning the rest
+        # TODO: make this less hard-code-y
+        n_snapshots, n_atoms = species.shape
+        max_dim = 160
+        output = torch.zeros((n_snapshots, n_atoms, max_dim)) * np.nan
+        # TODO: note intentional NaN-poisoning here -- not sure if there's a
+        #   better way to emulate jagged array
+
+        # loop through atom nets
+        for i, (_, module) in enumerate(self.items()):
+            mask = atom_species == i
+            # look only at the elements that are present in species
+            if sum(mask) > 0:
+                # get output for these atoms given the aev for these atoms
+                current_out = module(aev[:, mask, :])
+                # dimenstion of current_out is [nr_of_frames, nr_of_atoms_with_element_i,max_dim]
+                out_dim = current_out.shape[-1]
+                # jagged array
+                output[:, mask, :out_dim] = current_out
+                # final dimenstions are [n_snapshots, n_atoms, max_dim]
+
+        return SpeciesEnergies(species, output)
+
+
+class LastLayerANIModel(ANIModel):
+    """just like ANIModel, but only does the final calculation and cuts input arrays to the input feature size of the
+    different atom nets!"""
+
+    def forward(
+        self,
+        species_aev: Tuple[Tensor, Tensor],
+        cell: Optional[Tensor] = None,
+        pbc: Optional[Tensor] = None,
+    ) -> SpeciesEnergies:
+        species, aev = species_aev
+        species_ = species.flatten()
+        aev = aev.flatten(0, 1)
+
+        output = aev.new_zeros(species_.shape)
+
+        for i, (_, m) in enumerate(self.items()):
+            mask = species_ == i
+            midx = mask.nonzero().flatten()
+            if midx.shape[0] > 0:
+                input_ = aev.index_select(0, midx)
+                if i == 2:
```

Comment:
Nice! Minor tweak: rather than having an if/else block that covers element index i=2 and i=3, could we have an array (or dict) that maps `i` to `last_layer_dim`? Then we would say something like `input_ = input_[:, :last_layer_dim[i]]`

Reply:
yes, great point! I will update this!